crates/aptos/src/update/mod.rs

```diff
@@ -1,8 +1,91 @@
 // Copyright © Aptos Foundation
 // SPDX-License-Identifier: Apache-2.0
 
+// Note: We make use of the self_update crate, but as you can see in the case of
+// Revela, this can also be used to install / update other binaries.
+
+mod aptos;
 mod helpers;
+mod revela;
 mod tool;
 
-use helpers::check_if_update_required;
+use crate::common::types::CliTypedResult;
+use anyhow::{anyhow, Context, Result};
+pub use helpers::get_additional_binaries_dir;
+pub use revela::get_revela_path;
+use self_update::{update::ReleaseUpdate, version::bump_is_greater, Status};
 pub use tool::UpdateTool;
+
+/// Things that implement this trait are able to update a binary.
+trait BinaryUpdater {
+    /// Only used for messages we print to the user.
+    fn pretty_name(&self) -> &'static str;
+
+    /// Return information about whether an update is required.
+    fn get_update_info(&self) -> Result<UpdateRequiredInfo>;
+
+    /// Build the updater from the self_update crate.
+    fn build_updater(&self, info: &UpdateRequiredInfo) -> Result<Box<dyn ReleaseUpdate>>;
+
+    /// Update the binary. Install if not present, in the case of additional binaries
+    /// such as Revela.
+    fn update(&self) -> CliTypedResult<String> {
+        // Confirm that we need to update.
+        let info = self
+            .get_update_info()
+            .context("Failed to check if we need to update")?;
+        if !info.update_required()? {
+            return Ok(format!("Already up to date (v{})", info.target_version));
+        }
+
+        // Build the updater.
+        let updater = self.build_updater(&info)?;
+
+        // Update the binary.
+        let result = updater
+            .update()
+            .map_err(|e| anyhow!("Failed to update {}: {:#}", self.pretty_name(), e))?;
+
+        let message = match result {
+            Status::UpToDate(_) => unreachable!("We should have caught this already"),
+            Status::Updated(_) => match info.current_version {
+                Some(current_version) => format!(
+                    "Successfully updated {} from v{} to v{}",
```

Comment:
Should we print out like `program installed at 'path_to/prog', ensure it is in your PATH'? Or is it already guaranteed if the defaults are used? Not important, but just a thought.

Reply:
We do actually, here is the output from updating:
```
Checking target-arch... aarch64-apple-darwin
Checking current version... v1.0.0-rc3
Looking for tag: v1.0.0

revela release status:
  * New exe release: "revela-aarch64-apple-darwin"
  * New exe download url: "https://api.github.com/repos/verichains/revela/releases/assets/155171154"
  * Installing to: /Users/dport/.local/bin

The new release will be downloaded/extracted and the existing binary will be replaced.
Do you want to continue? [Y/n] y
Downloading...
Extracting archive... Moving binary file to /Users/dport/.local/bin... Done
{
  "Result": "Successfully updated Revela from v1.0.0-rc3 to v1.0.0"
}
```

See `Installing to: /Users/dport/.local/bin`.